src/Functions/array/arrayFold.cpp

```diff
@@ -0,0 +1,226 @@
+            const DataTypePtr & array_type_ptr = array_with_type_and_name.type;
+            const auto * array_type = checkAndGetDataType<DataTypeArray>(array_type_ptr.get());
+            if (!column_array)
+            {
+                const ColumnConst * column_const_array = checkAndGetColumnConst<ColumnArray>(column_array_ptr.get());
+                if (!column_const_array)
+                    throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Expected array column, found {}", column_array_ptr->getName());
+                column_array_ptr = recursiveRemoveLowCardinality(column_const_array->convertToFullColumn());
+                column_array = checkAndGetColumn<ColumnArray>(column_array_ptr.get());
+            }
+            if (!array_type)
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Expected array type, found {}", array_type_ptr->getName());
+            if (!offsets_column)
+            {
+                offsets_column = column_array->getOffsetsPtr();
+            }
+            else
+            {
+                /// The first condition is optimization: do not compare data if the pointers are equal.
+                if (column_array->getOffsetsPtr() != offsets_column
+                    && column_array->getOffsets() != typeid_cast<const ColumnArray::ColumnOffsets &>(*offsets_column).getData())
+                    throw Exception(ErrorCodes::SIZES_OF_ARRAYS_DONT_MATCH, "Arrays passed to {} must have equal size", getName());
+            }
+            if (i == 1)
+            {
+                column_first_array_ptr = column_array_ptr;
+                column_first_array = column_array;
+            }
+            arrays.emplace_back(ColumnWithTypeAndName(column_array->getDataPtr(),
+                                                      recursiveRemoveLowCardinality(array_type->getNestedType()),
+                                                      array_with_type_and_name.name));
+        }
+
+        ssize_t rows_count = input_rows_count;
+        ssize_t data_row_count = arrays[0].column->size();
+        auto array_count = arrays.size();
+
+        if (rows_count == 0)
+            return arguments.back().column->convertToFullColumnIfConst()->cloneEmpty();
+
+        ColumnPtr current_column;
+        current_column = arguments.back().column->convertToFullColumnIfConst();
+        MutableColumnPtr result_data = arguments.back().column->convertToFullColumnIfConst()->cloneEmpty();
+
+        size_t max_array_size = 0;
+        const auto & offsets = column_first_array->getOffsets();
+
+        //get columns of Nth array elements
+        IColumn::Selector selector(data_row_count);
+        size_t cur_ind = 0;
+        ssize_t cur_arr = 0;
+
+        if (data_row_count)
+            while (offsets[cur_arr] == 0)
+                ++cur_arr;
+
+        for (ssize_t i = 0; i < data_row_count; ++i)
+        {
+            selector[i] = cur_ind++;
```

Comment:
Not wrong, but for readability, let's make the postincrement a separate statement (that's also consistent with most of the remaining codebase).